Fix issues flagged by golangci-lint

dmitris opened the pull request:
PR fixes number of issues flagged by `golangci-lint`, mostly by the `gosimple` and `staticcheck` linters.

The issues flagged by golangci-lint (`golangci-lint run -D errcheck -D unused -D ineffassign ./...`):
https://gist.github.com/dmitris/b2b6836f2e90efa2131a74d8c5253ab7

Related to issue #3519.

SyntaxNode commented on adapters/adtelligent/adtelligent.go:
```diff
@@ -171,7 +171,8 @@ func validateImpression(imp *openrtb2.Imp) (int, error) {
 	var impExtBuffer []byte
 
-	impExtBuffer, err = json.Marshal(&adtelligentImpExt{
+	// TODO: confirm the missing error handling is intended
```
The error handling is not intentionally skipped. All json.Marshal errors should be bubbled up as a fatal error.

dmitris replied:
fixed in 146f9699

SyntaxNode commented on adapters/avocet/avocet.go:
```diff
@@ -85,6 +85,7 @@ func (a *AvocetAdapter) MakeBids(internalRequest *openrtb2.BidRequest, externalR
 			if len(br.SeatBid[i].Bid[j].Ext) > 0 {
 				err := json.Unmarshal(br.SeatBid[i].Bid[j].Ext, &ext)
 				if err != nil {
+					// TODO: find out why we are not using errs
```
This error should bubble up.

dmitris replied:
done in e382fe56

SyntaxNode commented on adapters/dmx/dmx.go:
```diff
@@ -103,7 +103,7 @@ func (adapter *DmxAdapter) MakeRequests(request *openrtb2.BidRequest, req *adapt
 			anyHasId = true
 		}
-		if anyHasId == false {
+		if !anyHasId {
```
I don't agree with this lint rule. Can we exclude it? It is arguably easier to see the "== false" than the "!" operator. Alternatively, can the functionality be changed such that this is hasNoID or missingID to better align with the usage?

dmitris replied:
reworked it to have `hasNoID` in a36d18ca00fe56181dc2b01e0a4bcaadcac0f5d5 - see if it looks ok to you @SyntaxNode

SyntaxNode commented on adapters/dmx/dmx.go:
```diff
@@ -263,8 +263,7 @@ func (adapter *DmxAdapter) MakeBids(request *openrtb2.BidRequest, externalReques
 func fetchParams(params dmxExt, inst openrtb2.Imp, ins openrtb2.Imp, imps []openrtb2.Imp, banner *openrtb2.Banner, video *openrtb2.Video, intVal int8) []openrtb2.Imp {
-	var tempimp openrtb2.Imp
-	tempimp = inst
+	var tempimp openrtb2.Imp = inst
```
We should prefer `tempimp := inst` here.

dmitris replied:
change in a36d18ca00fe56181dc2b01e0a4bcaadcac0f5d5 as well

SyntaxNode commented on adapters/gamma/gamma.go:
```diff
@@ -149,7 +149,7 @@ func (a *GammaAdapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *adapt
 		return nil, errs
 	}
-	var invalidImpIndex = make([]int, 0, 0)
+	var invalidImpIndex = make([]int, 0)
```
Let's use `var invalidImpIndex []int` instead to avoid the allocation when there are no invalid imps.

dmitris replied:
done in 25808d59

SyntaxNode commented on adapters/gamma/gamma.go:
```diff
@@ -204,8 +204,7 @@ func (a *GammaAdapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *adapt
 func convertBid(gBid gammaBid, mediaType openrtb_ext.BidType) *openrtb2.Bid {
-	var bid openrtb2.Bid
-	bid = gBid.Bid
+	var bid openrtb2.Bid = gBid.Bid
```
We should prefer `bid := gBid.Bid`.

SyntaxNode commented on adapters/lemmadigital/lemmadigital.go:
```diff
@@ -101,6 +101,8 @@ func (a *adapter) MakeBids(request *openrtb2.BidRequest, requestData *adapters.R
 			bidResponse.Bids = append(bidResponse.Bids, b)
 		}
+		// golangci-lint staticheck SA4004: the surrounding loop is unconditionally terminated
+		// TODO: confirm the break is intended and add '// nosemgrep: SA4004'
```
Please resolve this TODO.

dmitris replied:
done

SyntaxNode commented on adapters/minutemedia/minutemedia.go:
```diff
@@ -101,6 +101,8 @@ func extractOrg(openRTBRequest *openrtb2.BidRequest) (string, error) {
 		}
 
+		// golangci-lint staticcheck SA4004: the surrounding loop is unconditionally terminated
+		// TODO: confirm this is intended and add '// nosemgrep: SA4004'
```
Please resolve this TODO.

AlexBVolcy replied:
I see it’s labelled as resolved, but the TODO comment is still there. Can you address this?

dmitris replied:
done, added `//nolint: staticcheck // SA4004: the surrounding loop is unconditionally terminated`

SyntaxNode commented on adapters/smartrtb/smartrtb.go:
```diff
@@ -31,9 +31,9 @@ type bidRequestExt struct {
-	creativeTypeAudio         = "AUDIO"
+	creativeTypeVideo  string = "VIDEO"
+	creativeTypeNative string = "NATIVE"
+	creativeTypeAudio  string = "AUDIO"
```
This is valid Go code. Why does the linter want to change it?

dmitris replied:
here's the linter's issue description: [SA9004 - Only the first constant has an explicit type](https://staticcheck.io/docs/checks/#SA9004).

In this case, there is no risk of confusion about the constants type since it is `string`.  But if we think this way, why do we need to mark the first constant's type as `string` - and not the consequent once?  This is inconsistent and may lead to believe that the `string` type is propagate as it is when there is no value given.  But for now I will add a `//nolint` and restore the existing code.

dmitris replied:
reverted to the original version, added `// nolint: staticcheck // staticcheck SA9004: only the first constant in this group has an explicit type`

SyntaxNode commented on adapters/tappx/tappx_test.go:
```diff
@@ -38,13 +38,16 @@ func TestTsValue(t *testing.T) {
 
-	var test int
-	test = 0
+	var test int = 0
```
Surprised there's not a lint error for setting this to it's default value. var test int` or `test := 0` would be better.

dmitris replied:
done (changed to `test := 0`)

SyntaxNode commented on adapters/tappx/tappx_test.go:
```diff
@@ -38,13 +38,16 @@ func TestTsValue(t *testing.T) {
+	if err != nil {
+		t.Errorf("Error in buildEndpointURL: %s", err.Error())
+		return
+	}
```
Please use testify. This should be `require.NoError(t, err, "buildEndpointURL")`.

SyntaxNode commented on adservertargeting/respdataprocessor.go:
```diff
@@ -82,13 +82,14 @@ func buildBidExt(targetingData map[string]string,
 	newExt, err := jsonpatch.MergePatch(bid.Ext, bidExtTargeting)
 	if err != nil {
-		warnings = append(warnings, createWarning(err.Error()))
+		// warnings = append(warnings, createWarning(err.Error()))
```
Please resolve these TODOs.

dmitris replied:
removed `TODO`s and added the `//nolint: ineffassign,staticcheck` comment.  But it seems to me somewhere between fragile and wrong to append to slice passed by value - is it guaranteed that the caller will see those additions? https://go.dev/play/p/DmZumKNLFYC suggests it won't (or may not - depending on the capacity of the passed slice).  Should we change `buildBidExt` to take a pointer-to-slice `warnings *[]openrtb_ext.ExtBidderMessage` - or even return the `warnings` together with the `json.RawMessage`?

SyntaxNode commented on config/config.go:
```diff
@@ -182,32 +182,32 @@ func (data *ExternalCache) validate(errs []error) []error {
 
 	if data.Scheme != "" && data.Scheme != "http" && data.Scheme != "https" {
-		return append(errs, errors.New("External cache Scheme must be http or https if specified"))
+		return append(errs, errors.New("external cache Scheme must be http or https if specified"))
```
The case changes are potentially breaking, please revert.

dmitris replied:
done

SyntaxNode commented on floors/fetcher_test.go:
```diff
@@ -1219,7 +1219,6 @@ func TestPriceFloorFetcherSubmitFailed(t *testing.T) {
 }
 
 func getRandomNumber() int {
-	rand.Seed(time.Now().UnixNano())
```
What version of Go makes this unnecessary? We still support 1.19.

dmitris replied:
go.mod has  `go 1.20` so I thought it is the minimal supported version. I will revert and add a `//nolint` comment.

dmitris replied:
done

SyntaxNode commented on main.go:
```diff
@@ -80,7 +78,9 @@ func serve(cfg *config.Configuration) error {
 	corsRouter := router.SupportCORS(r)
-	server.Listen(cfg, router.NoCache{Handler: corsRouter}, router.Admin(currencyConverter, fetchingInterval), r.MetricsEngine)
+	if err := server.Listen(cfg, router.NoCache{Handler: corsRouter}, router.Admin(currencyConverter, fetchingInterval), r.MetricsEngine); err != nil {
+		glog.Errorf("prebid-server returned an error: %v", err)
```
Good catch on this error. I think this be a fatal "terminate now" error.

dmitris replied:
changing to `glog.Fatalf("prebid-server returned an error: %v", err)`

SyntaxNode commented on openrtb_ext/request_wrapper_test.go:
```diff
@@ -706,6 +706,7 @@ func TestCloneUserExt(t *testing.T) {
 				eids[1].UIDs[0].AType = 0
 				eids[0].UIDs = append(eids[0].UIDs, openrtb2.UID{ID: "Z", AType: 2})
 				eids = append(eids, openrtb2.EID{Source: "Blank"})
+				// TODO: double-check why we fill eids but don't use it
```
@hhhjort

dmitris replied:
(TODO(dmitris) - still open, wait for @hhhjort's response)

AlexBVolcy replied:
Can I get an update on this TODO please?

dmitris replied:
I admit I don't fully understand this test - but I added the nolint comment:
`eids = append(eids, openrtb2.EID{Source: "Blank"}) //nolint: ineffassign, staticcheck // this value of `eids` is never used (staticcheck)`
and remove TODO. /cc @hhhjort if you want to double-check it works as intended.

hhhjort replied:
Oh, yes, the clone tests. We start with a userext object and a manual copy of that object. (copy/paste of the code lines) We then make a clone of that object using the clone code we wish to test. Next we "mutate" the original, edit it first starting from the outermost reaches, working in towards the root trying to catch a change that the clone code does not protect against. Above we first edit the objects in the UIDs array, in case the clone array has pointers to the items. Then we append, which should replace the array itself, to make sure this doesn't somehow leak over to the clone.

In the end we make all sorts of changes to the original object. We don't care about those shanges, we aren't testing our ability to make these changes. We are only testing that the clone remains independent of all those changes, that the clone is still equal to the manual copy. Thus we get warnings that we don't use the edits we made to the original userext.

dmitris replied:
> Can I get an update on this TODO please?
@AlexBVolcy 

It has been removed, the code now:
```go
			mutator: func(t *testing.T, userExt *UserExt) {
				eids := *userExt.eids
				eids[0].UIDs[1].ID = "G2"
				eids[1].UIDs[0].AType = 0
				eids[0].UIDs = append(eids[0].UIDs, openrtb2.UID{ID: "Z", AType: 2})
				eids = append(eids, openrtb2.EID{Source: "Blank"}) //nolint: ineffassign, staticcheck // this value of `eids` is never used (staticcheck)
				userExt.eids = nil
			},
```

SyntaxNode commented on server/server_test.go:
```diff
@@ -75,7 +75,7 @@ func TestNewSocketServer(t *testing.T) {
 	assert.NotEqual(t, nil, ret, "ret : isNil()")
 	assert.Equal(t, mockServer.Addr, ret.Addr, fmt.Sprintf("Addr invalide: %v != %v",
-		ret.Addr, mockServer.Addr))
+		ret.Addr, mockServer.Addr)) //nolint:staticcheck // non-nil ret is ensures by the assert above
```
Why is this necessary for an assert statement?

dmitris replied:
here's the linter's output:
```bash
$ golangci-lint run ./... | grep server_test.go
server/server_test.go:77:39: SA5011: possible nil pointer dereference (staticcheck)
server/server_test.go:84:5: SA5011(related information): this check suggests that the pointer can be nil (staticcheck)
server/server_test.go:79:46: SA5011: possible nil pointer dereference (staticcheck)
server/server_test.go:114:5: SA5011(related information): this check suggests that the pointer can be nil (staticcheck)
```
There is a check `if ret != nil {` after those asserts which makes the linter suspicious.  I suggest we use `require.NotNil(t, ret, "ret : isNil()")` instead of `assert.NotEqual(t, nil, ret, "ret : isNil()")` and then we can delete the nil check around `ret.Close()`:
```go
       if ret != nil {
               ret.Close()
       }
```

dmitris replied:
removed these `//nolint` comments

AlexBVolcy commented on adapters/adtelligent/adtelligent.go:
```diff
@@ -171,7 +171,8 @@ func validateImpression(imp *openrtb2.Imp) (int, error) {
 
-	impExtBuffer, err = json.Marshal(&adtelligentImpExt{
+	// TODO: confirm the missing error handling is intended
+	impExtBuffer, _ /* err */ = json.Marshal(&adtelligentImpExt{
```
Just want to circle back to this to make sure the error is being properly handled and bubbled up, and that we can delete the TODO on line 174

dmitris replied:
ack - changed to:
```go
	impExtBuffer, err = json.Marshal(&adtelligentImpExt{
		Adtelligent: impExt,
	})
	if err != nil {
		return 0, &errortypes.BadInput{
			Message: fmt.Sprintf("ignoring imp id=%s, error while marshaling impExt, err: %s", imp.ID, err),
		}
	}
```

AlexBVolcy commented on adapters/rubicon/rubicon.go:
```diff
@@ -586,7 +586,7 @@ func createImpsToExtMap(imps []openrtb2.Imp) (map[*openrtb2.Imp]rubiconExtImpBid
 	preparedImpsToExtMap := make(map[*openrtb2.Imp]rubiconExtImpBidder)
 	for imp, bidderExt := range impsToExtMap {
-		if bidderExt.Bidder.BidOnMultiformat == false {
+		if !bidderExt.Bidder.BidOnMultiformat {
```
I agree with @SyntaxNode's other comments, where I prefer `bidderExt.Bidder.BidOnMultiformat == false` over using the `!` operator.

dmitris replied:
reverting back to the ` == false`. But I do prefer the Go idiomatic boolean check and consistency with the Go convention.  Even in this repository, this will be the only check of this type, while there are many cases of the standard form of checking for the boolean to be `false`:
```bash
$ rg -t go 'if !.* =='
exchange/utils.go
270:	if !defined || binfo.OpenRTB == nil {
960:	if !exists || fpdToApply == nil {

config/experiment.go
62:	if !(cfg.AdCerts.Mode == AdCertsSignerModeOff ||

exchange/events.go
84:	if !ev.isEventAllowed() || pbsBid.BidType == openrtb_ext.BidTypeVideo {
107:	if !ev.isEventAllowed() || pbsBid.BidType == openrtb_ext.BidTypeVideo {

endpoints/openrtb2/auction.go
682:		if !impExtBidderExists || impExtBidder == nil {
730:		if !impExtPrebidBidderExists || impExtPrebidBidder == nil {
1147:	if !hasRootSize && len(banner.Format) == 0 && !isInterstitial {

endpoints/openrtb2/auction_test.go
140:			if !info.IsDir() && filepath.Ext(info.Name()) == jsonFileExtension {

adapters/rubicon/rubicon.go
1062:			if !ok || bidCmpOverride == 0 {

endpoints/info/bidders.go
80:	if !exists || len(v) == 0 {

adapters/decenterads/decenterads.go
37:		if !bidderExtExists || len(bidderExt) == 0 {

adapters/bidscube/bidscube.go
38:		if !bidderExtExists || len(bidderExt) == 0 {
```

```bash
 rg -t go 'if .* == false'
adapters/rubicon/rubicon.go
590:		if bidderExt.Bidder.BidOnMultiformat == false { //nolint: staticcheck
```

AlexBVolcy commented on adapters/tappx/tappx.go:
```diff
@@ -133,6 +132,11 @@ func (a *TappxAdapter) buildEndpointURL(params *openrtb_ext.ExtImpTappx, test in
 
 	tappxHost := "tappx.com"
 	isNewEndpoint, err := regexp.Match(`^(zz|vz)[0-9]{3,}([a-z]{2,3}|test)$`, []byte(params.Endpoint))
+	if err != nil {
+		return "", &errortypes.BadInput{
+			Message: "Unable to match params.Endpoint " + string(params.Endpoint) + "): " + err.Error(),
+		}
+	}
```
Can you add a test to cover these lines?

dmitris replied:
TODO(dmitris): add a unit test

dmitris replied:
replied below https://github.com/prebid/prebid-server/pull/3621#issuecomment-2070923090 - not seeing an easy way to cover this, and suggest refactoring in a follow-up PR.

dmitris replied:
@AlexBVolcy - are you ok with leaving it as if for now?

AlexBVolcy replied:
Yes you can leave it as is, there are other uses of regexp.Match throughout the code base that also dont cover this error.

AlexBVolcy commented on adservertargeting/respdataprocessor.go:
```diff
@@ -82,13 +82,14 @@ func buildBidExt(targetingData map[string]string,
 	bidExtTargeting, err := jsonutil.Marshal(bidExtTargetingData)
 	if err != nil {
-		warnings = append(warnings, createWarning(err.Error()))
+		// TODO: figure out the intended usage of warnings, here and several lines below
```
Also noting that these TODOs and commented out code should be taken care of before approval.

dmitris replied:
removed `TODO`s and commented out code - but see above about assigning to a slice passed by value, I think it is wrong (or brittle).

AlexBVolcy commented on endpoints/events/vtrack.go:
```diff
@@ -271,12 +271,14 @@ func getBiddersAllowingVastUpdate(req *BidCacheRequest, bidderInfos *config.Bidd
-			return b.IsEnabled() && b.ModifyingVastXmlAllowed
+		if bidderInfos != nil {
+			if b, ok := (*bidderInfos)[normalizedBidder.String()]; ok {
+				// check if bidder is enabled
```
I don't think this comment is necessary with the code below explaining what it's doing well enough in my opinion.

dmitris replied:
👍

dmitris commented:
addressed most of the feedback, added a couple of `TODO(dmitris)` in the comments where I need to follow-up (make a test etc.).

@SyntaxNode do you want me to resolve the addressed comments or would you or the commenter resolve them when satisfied with the change?  Thanks.

AlexBVolcy commented on adapters/adtelligent/adtelligent.go:
```diff
@@ -175,6 +175,11 @@ func validateImpression(imp *openrtb2.Imp) (int, error) {
 	impExtBuffer, err = json.Marshal(&adtelligentImpExt{
 		Adtelligent: impExt,
 	})
+	if err != nil {
```
Can you add test for this new error?

AlexBVolcy replied:
Oh it's coming from `json.Marshal` than you can just ignore my comment then.

dmitris commented:
> For future reference, please don’t force push your changes as it makes the PR harder to review. Thank You!

So you don't want PRs rebased on the latest trunk version?  I usually ty to "rebase early and often" to ensure that the merged version wouldn't have any surprises, and to keep the PR commits on top of `git log`.

dmitris commented:
@AlexBVolcy - regarding [adapters/tappx/tappx.go](https://github.com/prebid/prebid-server/pull/3621/files#diff-6a270680d7b9688fa5c35e92df6f79573f0f4ec6b26262a89aedc179c8f2a927), `buildEndpointURL` - I don't see a way to "force" the regexp parsing error from the test file.  I think it should be refactored as here: https://github.com/dmitris/prebid-server/commit/c5af3039959bf135510ad091381346b21fff2fd9 - but would like to do it in a follow-up PR.  I suggest to leave this as is for now.

dmitris commented:
for reference: [these](https://gist.github.com/dmitris/67b4bf22e569f5eb2381994db0e179b1) are remaining issues flagged by `golangci-lint run -D errcheck -D unused -D ineffassign ./...` in the code version of this PR.

AlexBVolcy commented on adapters/dmx/dmx.go:
```diff
@@ -264,8 +264,7 @@ func (adapter *DmxAdapter) MakeBids(request *openrtb2.BidRequest, externalReques
 func fetchParams(params dmxExt, inst openrtb2.Imp, ins openrtb2.Imp, imps []openrtb2.Imp, banner *openrtb2.Banner, video *openrtb2.Video, intVal int8) []openrtb2.Imp {
-	var tempimp openrtb2.Imp
-	tempimp = inst
+	tempimp := inst
```
Nitpick: can you make this variable camel-case --> `tempImp`

AlexBVolcy commented on endpoints/openrtb2/auction.go:
```diff
@@ -2319,9 +2319,11 @@ func (deps *endpointDeps) processStoredRequests(requestJson []byte, impInfo []Im
-		_, err = jsonparser.ArrayEach(impArray, func(imp []byte, _ jsonparser.ValueType, _ int, err error) {
-			impExtData, _, _, err := jsonparser.Get(imp, "ext", "prebid")
+		_, _ = jsonparser.ArrayEach(impArray, func(imp []byte, _ jsonparser.ValueType, _ int, _ error) {
+			// ignore possible errors if there is no ext.prebid key
```
I think you can remove this comment

AlexBVolcy replied:
Just checking to see what you think about removing this

dmitris replied:
removed in 55345bfb

AlexBVolcy commented on endpoints/openrtb2/auction.go:
```diff
@@ -2319,9 +2319,11 @@ func (deps *endpointDeps) processStoredRequests(requestJson []byte, impInfo []Im
 
 // parseImpInfo parses the request JSON and returns impression and unmarshalled imp.ext.prebid
 func parseImpInfo(requestJson []byte) (impData []ImpExtPrebidData, errs []error) {
+	// ignore the error such as '&errors.errorString{s:"Unknown value type"}'
```
Should this comment be here? Since the line directly beneath does handle the error.

dmitris replied:
> Should this comment be here? Since the line directly beneath does handle the error.

But does it actually? The line underneath is:
`if impArray, dataType, _, err := jsonparser.Get(requestJson, "imp"); err == nil && dataType == jsonparser.Array {`
it doesn't handle cases where `err != nil` OR `dataType != jsonparser.Array`, right?  It just returns at the end of the function (L2337).  Please let me know if I'm missing something obvious 😄

AlexBVolcy replied:
Ah yes you're right, then do you mind me asking why did you add this comment? For my own understanding

AlexBVolcy commented on server/server_test.go:
```diff
@@ -181,6 +178,7 @@ func TestRunServer(t *testing.T) {
 }
 
 func TestListen(t *testing.T) {
+	// TODO(dmitris): check if the unused const 'name' can be deleted [PR3621]
```
Can I get an update on this TODO? I'd prefer to not merge in TODOs if possible.

dmitris replied:
I removed the unused variable and removed the `TODO` line in 5430f6c4.

AlexBVolcy replied:
Strange, I still see the TODO in the latest version of the code, could you double check this?

dmitris replied:
double-checked, no `TODO`s in `server/server_test.go` (or any files underneath `server/`) in the PR branch:
```bash
$ git log -1 && grep TODO server/server_test.go
commit ea94905c0a83e8c9db801f4b90a4e4a7524174da (HEAD -> golangci-lint, ds/golangci-lint)
Author: Dmitry S <dsavints@gmail.com>
Date:   Mon Apr 29 17:36:37 2024 +0200

    remove comment about error handling
$ rg TODO server/
$
```
[server/server_test.go](https://github.com/dmitris/prebid-server/blob/golangci-lint/server/server_test.go) in my / PR [branch](https://github.com/dmitris/prebid-server/blob/golangci-lint).

bsardo commented:
> > For future reference, please don’t force push your changes as it makes the PR harder to review. Thank You!
> 
> So you don't want PRs rebased on the latest trunk version? I usually ty to "rebase early and often" to ensure that the merged version wouldn't have any surprises, and to keep the PR commits on top of the commits list / `git log` output.

Hi @dmitris, we strongly prefer merging with master to rebasing once the review process has started because it makes it much easier for us to see what has changed based on comments. It can be a bit frustrating for reviewers to have to re-review everything every time something has changed. Even though commits merged in from master end up interspersed with your change commits, GitHub is smart enough to hide what has been merged in when viewing changes. We end up squashing everything anyway when we merge.

AlexBVolcy commented on endpoints/cookie_sync_test.go:
```diff
@@ -2221,8 +2221,8 @@ type FakeAccountsFetcher struct {
 
-func (f FakeAccountsFetcher) FetchAccount(ctx context.Context, defaultAccountJSON json.RawMessage, accountID string) (json.RawMessage, []error) {
-	defaultAccountJSON = json.RawMessage(`{"disabled":false}`)
+func (f FakeAccountsFetcher) FetchAccount(ctx context.Context, _ json.RawMessage, accountID string) (json.RawMessage, []error) {
```
Can't we get rid of the `_ json.RawMessage` altogether now that you instantiate the variable below?

AlexBVolcy reviewed (commented):
Left some more comments, I'm also going to leave one more regarding the output I get when running the golangci-lint command on this PR

AlexBVolcy commented:
So when I run this command: golangci-lint run --disable-all -E staticcheck ./...

I get this output:

```
alexvolcy@vpn-10-249-98-42 prebid-server % golangci-lint run --disable-all -E staticcheck ./...
adapters/yandex/yandex.go:159:40: SA1019: yandexExt.ImpID is deprecated: in favor of `PlacementID` (staticcheck)
                placementID.ImpID = strconv.Itoa(int(yandexExt.ImpID))
                                                     ^
adapters/yandex/yandex.go:160:41: SA1019: yandexExt.PageID is deprecated: in favor of `PlacementID` (staticcheck)
                placementID.PageID = strconv.Itoa(int(yandexExt.PageID))
                                                      ^
server/ssl/ssl_test.go:15:14: SA1019: certPool.Subjects has been deprecated since Go 1.18: if s was returned by [SystemCertPool], Subjects will not include the system roots. (staticcheck)
        subjects := certPool.Subjects()
                    ^
server/ssl/ssl_test.go:25:13: SA1019: certPool.Subjects has been deprecated since Go 1.18: if s was returned by [SystemCertPool], Subjects will not include the system roots. (staticcheck)
        subjects = certPool.Subjects()
                   ^
server/ssl/ssl_test.go:40:14: SA1019: certPool.Subjects has been deprecated since Go 1.18: if s was returned by [SystemCertPool], Subjects will not include the system roots. (staticcheck)
        subjects := certPool.Subjects()
                    ^
adapters/pulsepoint/pulsepoint.go:142:7: SA4022: the address of a variable cannot be nil (staticcheck)
                        if &imp != nil && bidType != "" {
                           ^
usersync/cookie_test.go:649:12: SA4001: *&x will be simplified to x. It will not copy x. (staticcheck)
        cookie := *(&Cookie{
                  ^
adapters/adtarget/adtarget.go:174:2: SA4006: this value of `err` is never used (staticcheck)
        impExtBuffer, err = json.Marshal(&adtargetImpExt{
        ^
adapters/algorix/algorix.go:144:5: SA4006: this value of `err` is never used (staticcheck)
                                videoCopy.Ext, err = json.Marshal(&videoExt)
                                ^
adapters/grid/grid.go:397:4: SA4006: this value of `err` is never used (staticcheck)
                        bidMeta, err := getBidMeta(sb.Bid[i].Ext)
                        ^
hooks/hookexecution/execution.go:346:3: SA4006: this value of `ipConf` is never used (staticcheck)
                ipConf := privacy.IPConf{}
                ^
hooks/hookexecution/executor.go:205:2: SA4006: this value of `payload` is never used (staticcheck)
        outcome, payload, contexts, reject := executeStage(executionCtx, plan, payload, handler, e.metricEngine)
        ^
openrtb_ext/request_wrapper_test.go:708:5: SA4006: this value of `eids` is never used (staticcheck)
                                eids = append(eids, openrtb2.EID{Source: "Blank"})
                                ^
adapters/adgeneration/adgeneration_test.go:262:51: SA5011: possible nil pointer dereference (staticcheck)
        assert.Equal(t, expectedCurrency, bidderResponse.Currency)
                                                         ^
adapters/adgeneration/adgeneration_test.go:248:5: SA5011(related information): this check suggests that the pointer can be nil (staticcheck)
        if bidderResponse == nil {
           ^
adapters/adgeneration/adgeneration_test.go:263:40: SA5011: possible nil pointer dereference (staticcheck)
        assert.Equal(t, 1, len(bidderResponse.Bids))
                                              ^
adapters/adgeneration/adgeneration_test.go:264:45: SA5011: possible nil pointer dereference (staticcheck)
        assert.Equal(t, expectedID, bidderResponse.Bids[0].Bid.ID)
                                                   ^
adapters/telaria/telaria.go:207:5: SA5011(related information): this check suggests that the pointer can be nil (staticcheck)
        if telariaImpExt != nil && telariaImpExt.Extra != nil {
           ^
exchange/bidder_test.go:1068:27: SA5011(related information): this check suggests that the pointer can be nil (staticcheck)
                assert.Equal(t, false, (seatBid == nil && tc.expectedBidsCount != 0), tc.description)
                                        ^
adapters/consumable/consumable.go:97:14: SA4010: this result of append is never used, except maybe in other appends (staticcheck)
                                errors = append(errors, err)
                                         ^
hooks/hookexecution/mocks_test.go:61:8: SA4009: argument payload is overwritten before first use (staticcheck)
                func(payload hookstage.RawAuctionRequestPayload) (hookstage.RawAuctionRequestPayload, error) {
                     ^
hooks/hookexecution/mocks_test.go:62:4: SA4009(related information): assignment to payload (staticcheck)
                        payload = []byte(`{"name": "John", "last_name": "Doe", "foo": "bar"}`)
                        ^
hooks/hookexecution/mocks_test.go:66:8: SA4009: argument payload is overwritten before first use (staticcheck)
                func(payload hookstage.RawAuctionRequestPayload) (hookstage.RawAuctionRequestPayload, error) {
                     ^
hooks/hookexecution/mocks_test.go:67:4: SA4009(related information): assignment to payload (staticcheck)
                        payload = []byte(`{"last_name": "Doe", "foo": "bar"}`)
                        ^
hooks/hookexecution/mocks_test.go:123:21: SA4009: argument payload is overwritten before first use (staticcheck)
        c.AddMutation(func(payload hookstage.RawAuctionRequestPayload) (hookstage.RawAuctionRequestPayload, error) {
                           ^
hooks/hookexecution/mocks_test.go:124:3: SA4009(related information): assignment to payload (staticcheck)
                payload = []byte(`{"last_name": "Doe", "foo": "bar", "address": "A st."}`)
                ^
exchange/auction.go:47:20: SA1006: printf-style function with dynamic format string and no further arguments should use print-style function instead (staticcheck)
                d.Data.Request = fmt.Sprintf(d.Regexp.ReplaceAllString(d.Data.Request, ""))
                                 ^
exchange/auction.go:48:20: SA1006: printf-style function with dynamic format string and no further arguments should use print-style function instead (staticcheck)
                d.Data.Headers = fmt.Sprintf(d.Regexp.ReplaceAllString(d.Data.Headers, ""))
                                 ^
exchange/auction.go:49:21: SA1006: printf-style function with dynamic format string and no further arguments should use print-style function instead (staticcheck)
                d.Data.Response = fmt.Sprintf(d.Regexp.ReplaceAllString(d.Data.Response, ""))
                                  ^
exchange/exchange_test.go:2598:68: SA1012: do not pass a nil Context, even if a function permits it; pass context.TODO if you are unsure about which Context to use (staticcheck)
        bidCategory, adapterBids, rejections, err := applyCategoryMapping(nil, *requestExt.Prebid.Targeting, adapterBids, categoriesFetcher, targData, &randomDeduplicateBidBooleanGenerator{}, &nonBids{})
                                                                          ^
exchange/exchange_test.go:2653:68: SA1012: do not pass a nil Context, even if a function permits it; pass context.TODO if you are unsure about which Context to use (staticcheck)
        bidCategory, adapterBids, rejections, err := applyCategoryMapping(nil, *requestExt.Prebid.Targeting, adapterBids, categoriesFetcher, targData, &randomDeduplicateBidBooleanGenerator{}, &nonBids{})
                                                                          ^
exchange/exchange_test.go:2705:68: SA1012: do not pass a nil Context, even if a function permits it; pass context.TODO if you are unsure about which Context to use (staticcheck)
        bidCategory, adapterBids, rejections, err := applyCategoryMapping(nil, *requestExt.Prebid.Targeting, adapterBids, categoriesFetcher, targData, &randomDeduplicateBidBooleanGenerator{}, &nonBids{})
                                                                          ^
stored_requests/events/http/http.go:83:15: SA1015: using time.Tick leaks the underlying ticker, consider using it only in endless functions, tests and the main package, and use time.NewTicker here (staticcheck)
        go e.refresh(time.Tick(refreshRate))
```

For my understanding, is there a reason why some of these still appear and weren't addressed in this PR?

Thank you :)

dmitris commented:
@AlexBVolcy - the main reason is to limit the scope / size of the PR; my idea was to get this through and then used the experience / reviewer's comments for the "next batch" / follow-up.

dmitris commented:
@AlexBVolcy I believe I removed all the `TODO`s from the PR changes.  Please let me know if there are anything else we need to resolve before we can proceed to merge the PR.  Don't know why the [validate-merge](https://github.com/prebid/prebid-server/actions/runs/8881372865/job/24383551442?pr=3621) job failed - something is flaky? could you retry, please?  Locally (on a Mac), `go test ./...` and `./validate.sh --nofmt --cov --race 10` seem to work fine, repeatedly.

AlexBVolcy commented:
@dmitris Thank you for your work on this PR. Yes you're right about the flaky test failure, we have fix coming in for that soon so don't worry. I'm giving this PR another review now, so I'll have some more feedback soon. Thank you!

AlexBVolcy commented:
@dmitris This is looking good, I'm going to bring up in a team meeting whether the remaining issues that are being flagged by `golang-ci` should be also addressed in this PR, or in a different one like you suggested.

AlexBVolcy commented:
@dmitris Discussed with the team and confirmed that it's okay if we address the remaining `golanci-lint` issues in future "next batch" PRs.

guscarreon commented on adapters/lemmadigital/lemmadigital.go:
```diff
@@ -102,7 +102,7 @@ func (a *adapter) MakeBids(request *openrtb2.BidRequest, requestData *adapters.R
 			bidResponse.Bids = append(bidResponse.Bids, b)
 		}
-		break
+		break //nolint: staticcheck // break is intended, exempt from staticcheck SA4004
```
Instead of the `nolint` tag, can we remove the `break` entirely?
```
 97 -     for _, seatBid := range response.SeatBid {
 98 -         for i := range seatBid.Bid {
    +     if len(response.SeatBid) > 0 {
    +         for i := range response.SeatBid[0].Bid {
 99               b := &adapters.TypedBid{
100 -                 Bid:     &seatBid.Bid[i],
    +                 Bid:     &response.SeatBid[0].Bid[i],
101                   BidType: bidType,
102               }
103               bidResponse.Bids = append(bidResponse.Bids, b)
104           }
105 -         break //nolint: staticcheck // break is intended, exempt from staticcheck SA400
106       }
107  
108       return bidResponse, nil
109   }
adapters/lemmadigital/lemmadigital.go [RO]
```

dmitris replied:
done in 6d4f03aa

guscarreon replied:
Thank you

guscarreon commented on adapters/minutemedia/minutemedia.go:
```diff
@@ -102,6 +102,7 @@ func extractOrg(openRTBRequest *openrtb2.BidRequest) (string, error) {
 		}
 
+		//nolint: staticcheck // SA4004: the surrounding loop is unconditionally terminated
 		return strings.TrimSpace(impExt.Org), nil
```
Same here, I believe we can refactor to avoid the `nolint` comment:
```
 92   func extractOrg(openRTBRequest *openrtb2.BidRequest) (string, error) {
 93       var err error
 94 -     for _, imp := range openRTBRequest.Imp {
    +     if len(openRTBRequest.Imp) == 0 {
    +         return "", errors.New("no imps in bid request")
    +     }
 95           var bidderExt adapters.ExtImpBidder
 96           if err = json.Unmarshal(imp.Ext, &bidderExt); err != nil {
 97               return "", fmt.Errorf("failed to unmarshal bidderExt: %w", err)
 98           }
 99  
100           var impExt openrtb_ext.ImpExtMinuteMedia
101           if err = json.Unmarshal(bidderExt.Bidder, &impExt); err != nil {
102               return "", fmt.Errorf("failed to unmarshal ImpExtMinuteMedia: %w", err)
103           }
104  
105 -         //nolint: staticcheck // SA4004: the surrounding loop is unconditionally terminated
106 -         return strings.TrimSpace(impExt.Org), nil
107 -     }
108 -
109 -     return "", errors.New("no imps in bid request")
    +     return strings.TrimSpace(impExt.Org), nil
110   }
adapters/minutemedia/minutemedia.go [RO]
```

dmitris replied:
done in 6d4f03aa

guscarreon replied:
Thank you

guscarreon commented on adapters/tappx/tappx_test.go:
```diff
@@ -38,13 +39,16 @@ func TestTsValue(t *testing.T) {
 
 	url, err := bidderTappx.buildEndpointURL(&tappxExt, test)
+	if err != nil {
+		require.NoError(t, err, "buildEndpointURL")
```
No need for the `return` statement because `require.NoError` stops the execution of `TestTsValue` if it fails. Can we remove?
```
47     url, err := bidderTappx.buildEndpointURL(&tappxExt, test)
48 -   if err != nil {
49         require.NoError(t, err, "buildEndpointURL")
50 -       return
51 -   }
adapters/tappx/tappx_test.go [RO]
```

dmitris replied:
changed in 6d4f03aa

guscarreon replied:
> Why don't we use require.NoError in the next lines, after match, err := regexp.MatchString? Should I change it as well (can do in a follow-up PR)?

I completely agree. Sure, it's better to do in a follow up PR. Thanks for the change.

guscarreon commented on adapters/yieldmo/yieldmo.go:
```diff
@@ -35,14 +35,12 @@ type ExtBid struct {
 	if adapterReq != nil {
 		adapterRequests = append(adapterRequests, adapterReq)
 	}
-	errs = append(errs, errors...)
```
If we remove `var errs []error`, it looks odd to not remove `var adapterRequests []*adapters.RequestData` for the same reason. Do you agree?

Maybe we can simplify to:
```
37   func (a *YieldmoAdapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
38 -     var adapterRequests []*adapters.RequestData
39 -
40 -     adapterReq, errors := a.makeRequest(request)
41 -     if adapterReq != nil {
42 -         adapterRequests = append(adapterRequests, adapterReq)
43 -     }
44 -
45 -     return adapterRequests, errors
   +     return a.makeRequest(request)
46   }
adapters/yieldmo/yieldmo.go
```
Whatsmore, we could even rename `makeRequest` to `MakeRequests` and remove an extra function entirely. Do you think this would be outside the scope of this PR?
```
37 - func (a *YieldmoAdapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
38 -     var adapterRequests []*adapters.RequestData
39 -
40 -     adapterReq, errors := a.makeRequest(request)
41 -     if adapterReq != nil {
42 -         adapterRequests = append(adapterRequests, adapterReq)
43 -     }
44 -
45 -     return adapterRequests, errors
46 - }
47  
48 - func (a *YieldmoAdapter) makeRequest(request *openrtb2.BidRequest) (*adapters.RequestData, []error) {
   + func (a *YieldmoAdapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
49       var errs []error
50  
51       if err := preprocess(request); err != nil {
52           errs = append(errs, err)
53       }
54  
55       // Last Step
56       reqJSON, err := json.Marshal(request)
57       if err != nil {
58           errs = append(errs, err)
adapters/yieldmo/yieldmo.go
```

dmitris replied:
I don't see a way to simplify `YieldmoAdapter.MakeRequests()` to a one-liner since  `a.makeRequest(request)` returns a single `*adapters.RequestData` pointer, and the `MakeRequests` method needs to return a slice of such pointers, `[]*adapters.RequestData`.   But we can have a simplified two-line version:
```go
func (a *YieldmoAdapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
	reqData, errors := a.makeRequest(request)
	return []*adapters.RequestData{reqData}, errors
}
```

Regarding other refactoring, I would prefer to do it in a separate / follow-up PR

dmitris replied:
changed in 6d4f03aa

guscarreon replied:
> I don't see a way to simplify `YieldmoAdapter.MakeRequests()` to a one-liner since `a.makeRequest(request)` returns a single `*adapters.RequestData` pointer

You're right @dmitris. I didn't notice that. Thanks for the change.

guscarreon commented on endpoints/openrtb2/auction.go:
```diff
@@ -2320,8 +2320,8 @@ func (deps *endpointDeps) processStoredRequests(requestJson []byte, impInfo []Im
-		_, err = jsonparser.ArrayEach(impArray, func(imp []byte, _ jsonparser.ValueType, _ int, err error) {
-			impExtData, _, _, err := jsonparser.Get(imp, "ext", "prebid")
+		_, _ = jsonparser.ArrayEach(impArray, func(imp []byte, _ jsonparser.ValueType, _ int, _ error) {
+			impExtData, _, _, _ := jsonparser.Get(imp, "ext", "prebid")
```
Instead of ignoring possible errors in `jsonparser.ArrayEach` and `jsonparser.Get`, can we add to the array and return instead?

dmitris replied:
I'm trying to do it (as below) - but then multiple tests like `TestJsonSampleRequests/Asserts_we_return_400s_on_requests_that_are_not_supposed_to_pass_validation#25` start to fail.  Maybe we could do it in a follow-up PR?
```go
// parseImpInfo parses the request JSON and returns impression and unmarshalled imp.ext.prebid
func parseImpInfo(requestJson []byte) (impData []ImpExtPrebidData, errs []error) {
	impArray, dataType, _, err := jsonparser.Get(requestJson, "imp")
	if err != nil {
		errs = append(errs, err)
		return
	}
	if dataType == jsonparser.Array {
		_, err = jsonparser.ArrayEach(impArray, func(imp []byte, _ jsonparser.ValueType, _ int, _ error) {
			if err != nil {
				// errs = append(errs, err)
			}
			impExtData, _, _, _ := jsonparser.Get(imp, "ext", "prebid")
			var impExtPrebid openrtb_ext.ExtImpPrebid
			if impExtData != nil {
				if err := jsonutil.Unmarshal(impExtData, &impExtPrebid); err != nil {
					errs = append(errs, err)
				}
			}
			newImpData := ImpExtPrebidData{imp, impExtPrebid}
			impData = append(impData, newImpData)
		})
	}
	return
}
```

guscarreon replied:
> Maybe we could do it in a follow-up PR?

Sure @dmitris. Thank you for looking into this

dmitris commented:
@guscarreon - addressed your comments (except the last one, see my reply) in 6d4f03aa - could you double-check and resolve the conversations, please?

guscarreon reviewed (approved):
LGTM. Thank you @dmitris  for addressing our feedback
